build-support/rust: Fix sysroot for cross

Ericson2314 opened the pull request:
###### Motivation for this change

FIx the sysroot used for rust cross after the Rust src layout changed. Also split things out in preparation for eventually using this stuff with `buildRustCrate` too.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Ericson2314 committed:
build-support/rust: Split out sysroot src derivation

Hoping to make it usable for `buildRustCrate` too.

Ericson2314 committed:
rustcSrc: Reduce duplication

Ericson2314 committed:
Update script as rust-src layout has changed

Use stub lib so `core` and `alloc` are handled symmetrically.

SuperSandro2000 commented on pkgs/development/compilers/rust/rust-src.nix:
```diff
@@ -1,14 +1,25 @@
-      then "ci,doc,etc,grammar,llvm-project,llvm-emscripten,rtstartup,rustllvm,test,tools,vendor,stdarch"
-      else "ci,doc,etc,grammar,llvm-project,llvm-emscripten,rtstartup,rustllvm,test,vendor"
-    }}
+    rm -rf $out/{${lib.concatStringsSep "," ([
```
```suggestion
    rm -r $out/{${lib.concatStringsSep "," ([
```
Should it fail if any of the directories changes?

Ericson2314 replied:
@SuperSandro2000 I don't think this list is up to date since the src was reorganized, so I rather deal with that later. Indeed for the rest of the PR, I use `rustLibSrc` instead.

SuperSandro2000 commented on pkgs/build-support/rust/sysroot/update-lockfile.sh:
```diff
@@ -0,0 +1,27 @@
+NIXPKGS_ROOT="$HERE/../../../.."
+
+# https://unix.stackexchange.com/a/84980/390173
+tempdir=$(mktemp -d 2>/dev/null || mktemp -d -t 'update-lockfile')
```
Is this not using nix' coreutils? Maybe we need to add them to the packages for the shell?

Ericson2314 replied:
N.B. this was in the script before I moved it so I soon as well deal with later.

SuperSandro2000 commented on pkgs/build-support/rust/sysroot/update-lockfile.sh:
```diff
@@ -0,0 +1,27 @@
+
+cp Cargo.lock "$HERE"
+
+rm -rf "$tempdir"
```
This should probably be also run when the script fails.

Ericson2314 replied:
I actually I think it is good if it it left behind to debug of something fails.

Remember this is an ad-hoc maintainer script.

SuperSandro2000 replied:
Fine for me.

SuperSandro2000 commented on pkgs/build-support/rust/sysroot/src.nix:
```diff
@@ -0,0 +1,22 @@
+stdenv.mkDerivation {
+  name = "cargo-src";
+  preferLocalBuild = true;
+  phases = [ "installPhase" ];
```
```suggestion
```
Please use dont* equivalents.

Ericson2314 committed:
build-support/rust/sysroot/src: Use `dont*` instead of phase list

Making this separate commit because the original was moved out just the
way it was done before.
